File: CRUSH/7-19/linked-list/linked-list.cpp
/* Corey Garvey - July 19
 * Linked List */
#include <iostream>

using namespace std;

struct Node{
    int data;
    Node* next;
};

class LinkedList{
    public:
        Node* head;
        LinkedList(int val){
            Node* newNode = new Node;
            newNode->data = val;
            newNode->next = nullptr;
            head = newNode;
        }
        void appendToTail(int val){
            Node* current = head;
            while(current->next != nullptr){
                current = current->next;
            }
            Node* tmp = new Node;
            tmp->data = val;
            tmp->next = nullptr;
            current->next = tmp;
        }
        void appendToHead(int val){
            Node* tmp = new Node;
            tmp->data = val;
            tmp->next = head;
            head = tmp;
        }
        Node* removeNode(int val){
            Node* current = head;
            if(current->data == val){
                head = current->next;
                delete current;
            }
            while(current->next != nullptr){
                if(current->next->data == val){
                    Node* tmp = current->next;
                    current->next = current->next->next;
                    cout << "found it" << endl;
                    return tmp;
                }
                current = current->next;
            }
            cout << "Sorry, not in list" << endl;
            return nullptr;
        }
};
int main(int argc, char** argv){
    LinkedList myList = LinkedList(1);
    myList.appendToTail(2);
    myList.appendToTail(3);
    myList.appendToHead(4);
    myList.removeNode(1);
    Node* head = myList.head;
    cout << head->data << endl;
    cout << head->next->data << endl;
    cout << head->next->next->data << endl;
    return 0; 
}

File: python/linkedLists/linkedList.py
'''

Singly linked list in Python

'''

import sys

class Node:

	def __init__(self, data=None, next_node=None):

		self.data = data

		self.next_node = next_node


	def get_data(self):

		return self.data


class LinkedList:

	def __init__(self, head=None):

		self.head = head

	def insert(self, data=None):

		newNode = Node(data)

		current = self.head

		while current.next_node != None:

			current = current.next_node

		current.next_node = newNode

		return

	def pop(self):

		current = self.head

		while current.next_node.next_node != None:

			current = current.next_node

		tmp = current.next_node.next_node

		del current.next_node

		current.next_node = tmp

		return


	def deleteNode(self, data):

		current = self.head

		if(current.data == data):

			tmp = current.next_node

			del current

			self.head = tmp

			print "Deleted your node, here's the new LL: "

			self.printLL()

			return

		while(current.next_node != None):

			if(current.next_node.data == data):

				tmp = current.next_node.next_node

				del current.next_node

				current.next_node = tmp

				print "Deleted your node, here's the new LL: "

				self.printLL()

				return

			current = current.next_node

		print "Sorry, no node with that data"

		return

	
	def get_length(self):

		count = 0

		current = self.head

		while current != None:

			current = current.next_node

			count += 1

		return count


	def printLL(self):

		current = self.head

		outputString = ""

		while current.next_node != None:

			outputString += str(current.data) + "-"

			current = current.next_node

		outputString += str(current.data)

		print outputString

		return


#if __name__ == "__main__":
def main():

	filename = sys.argv[0]

	myNode = Node(1)

	myLL = LinkedList(myNode)

	myLL.insert(2)

	myLL.insert(3)

	myLL.insert(4)

	myLL.insert(5)

	myLL.printLL()

	myLL.pop()

	myLL.insert(6)

	myLL.printLL()

	deleteMe = int(raw_input("Enter a node to delete: "))

	myLL.deleteNode(deleteMe)

	inputNode = raw_input("Enter a node to input (None to exit): ")

	while inputNode != "None":

		myLL.insert(inputNode)

		print "Your new LinkedList: "

		myLL.printLL()

		inputNode = raw_input("Enter a node to input (None to exit): ")


main()